File: packages/crypto/src/encoding.ts
import { base58 } from "@scure/base";

/**
 * Multicodec prefix for Ed25519 public keys.
 * varint 0xed = [0xed, 0x01]
 */
export const ED25519_MULTICODEC_PREFIX = new Uint8Array([0xed, 0x01]);

/**
 * Multicodec prefix for Ed25519 private keys.
 * varint 0x1300 = [0x80, 0x26]
 */
export const ED25519_PRIV_MULTICODEC_PREFIX = new Uint8Array([0x80, 0x26]);

/**
 * Encode bytes as a multibase base58btc string.
 * Prefix 'z' indicates base58btc encoding per the multibase spec.
 * @param bytes - The bytes to encode.
 * @returns A multibase string starting with 'z'.
 */
export function encodeMultibase(bytes: Uint8Array): string {
  return "z" + base58.encode(bytes);
}

/**
 * Decode a multibase base58btc string to bytes.
 * @param encoded - A multibase string starting with 'z'.
 * @returns The decoded bytes.
 * @throws If the string is empty, does not start with 'z', or decoding fails.
 */
export function decodeMultibase(encoded: string): Uint8Array {
  if (encoded.length === 0) {
    throw new Error("Empty input: missing multibase prefix.");
  }
  if (!encoded.startsWith("z")) {
    throw new Error(
      `Unsupported multibase prefix: '${encoded[0]}'. Expected 'z' (base58btc).`,
    );
  }
  return base58.decode(encoded.slice(1));
}

/**
 * Decode a multibase-encoded Ed25519 public key to 32 raw bytes.
 * Strips the multicodec prefix (0xed 0x01) if present.
 *
 * @param encoded - Multibase string (e.g. from a DID or request body).
 * @returns 32-byte Ed25519 public key.
 * @throws If decoding fails or the result is not 32 bytes.
 */
export function decodePublicKey(encoded: string): Uint8Array {
  const bytes = decodeMultibase(encoded);
  let raw: Uint8Array = bytes;
  if (
    bytes.length === 34 &&
    bytes[0] === ED25519_MULTICODEC_PREFIX[0] &&
    bytes[1] === ED25519_MULTICODEC_PREFIX[1]
  ) {
    raw = bytes.slice(2);
  }
  if (raw.length !== 32) {
    throw new Error(
      `Invalid public key length: expected 32 bytes (Ed25519), got ${raw.length} after decoding.`,
    );
  }
  return raw;
}

/**
 * Decode a multibase-encoded Ed25519 private key to 32 raw bytes.
 * Strips the multicodec prefix (0x80 0x26 / 0x1300) if present.
 *
 * @param encoded - Multibase string (e.g. from identity storage).
 * @returns 32-byte Ed25519 private key.
 * @throws If decoding fails or the result is not 32 bytes.
 */
export function decodePrivateKey(encoded: string): Uint8Array {
  const bytes = decodeMultibase(encoded);
  let raw: Uint8Array = bytes;
  if (
    bytes.length === 34 &&
    bytes[0] === ED25519_PRIV_MULTICODEC_PREFIX[0] &&
    bytes[1] === ED25519_PRIV_MULTICODEC_PREFIX[1]
  ) {
    raw = bytes.slice(2);
  }
  if (raw.length !== 32) {
    throw new Error(
      `Invalid private key length: expected 32 bytes (Ed25519), got ${raw.length} after decoding.`,
    );
  }
  return raw;
}

/**
 * Encode a 32-byte Ed25519 private key seed as multibase with multicodec prefix.
 *
 * @param raw - 32-byte Ed25519 private key seed.
 * @returns Multibase string with Ed25519 private key multicodec prefix.
 */
export function encodePrivateKey(raw: Uint8Array): string {
  if (raw.length !== 32) {
    throw new Error(`Expected 32-byte Ed25519 key, got ${raw.length} bytes.`);
  }
  const prefixed = new Uint8Array(
    ED25519_PRIV_MULTICODEC_PREFIX.length + raw.length,
  );
  prefixed.set(ED25519_PRIV_MULTICODEC_PREFIX);
  prefixed.set(raw, ED25519_PRIV_MULTICODEC_PREFIX.length);
  return encodeMultibase(prefixed);
}

/**
 * Derive a did:syr identifier from an Ed25519 public key.
 *
 * Process:
 * 1. Prepend multicodec prefix for Ed25519 public key (0xed01)
 * 2. Encode as base58btc multibase (prefix 'z')
 * 3. Prepend 'did:syr:'
 *
 * @param publicKey - The Ed25519 public key (32 bytes).
 * @returns The did:syr identifier string.
 */
export function deriveDid(publicKey: Uint8Array): string {
  if (publicKey.length !== 32) {
    throw new Error(
      `Expected 32-byte Ed25519 public key, got ${publicKey.length} bytes.`,
    );
  }

  // Prepend multicodec prefix
  const prefixed = new Uint8Array(
    ED25519_MULTICODEC_PREFIX.length + publicKey.length,
  );
  prefixed.set(ED25519_MULTICODEC_PREFIX);
  prefixed.set(publicKey, ED25519_MULTICODEC_PREFIX.length);

  // Encode as multibase base58btc
  const multibase = encodeMultibase(prefixed);

  return `did:syr:${multibase}`;
}

File: packages/rust/syr-crypto-core/src/encoding.rs
//! Multibase encoding/decoding and DID derivation.

use bs58;

/// Multicodec prefix for Ed25519 public keys. varint 0xed = [0xed, 0x01]
pub const ED25519_MULTICODEC_PREFIX: [u8; 2] = [0xed, 0x01];

/// Multicodec prefix for Ed25519 private keys. varint 0x1300 = [0x80, 0x26]
pub const ED25519_PRIV_MULTICODEC_PREFIX: [u8; 2] = [0x80, 0x26];

/// Encode bytes as multibase base58btc string (prefix 'z').
pub fn encode_multibase(bytes: &[u8]) -> String {
    format!("z{}", bs58::encode(bytes).into_string())
}

/// Decode multibase base58btc string to bytes.
pub fn decode_multibase(encoded: &str) -> Result<Vec<u8>, String> {
    if encoded.is_empty() {
        return Err("Empty input: missing multibase prefix.".to_string());
    }
    if !encoded.starts_with('z') {
        return Err(format!(
            "Unsupported multibase prefix: '{}'. Expected 'z' (base58btc).",
            encoded.chars().next().unwrap_or(' ')
        ));
    }
    bs58::decode(&encoded[1..])
        .into_vec()
        .map_err(|e| e.to_string())
}

/// Decode multibase-encoded Ed25519 public key to 32 raw bytes.
pub fn decode_public_key(encoded: &str) -> Result<[u8; 32], String> {
    let bytes = decode_multibase(encoded)?;
    let raw: Vec<u8> = if bytes.len() == 34
        && bytes[0] == ED25519_MULTICODEC_PREFIX[0]
        && bytes[1] == ED25519_MULTICODEC_PREFIX[1]
    {
        bytes[2..].to_vec()
    } else {
        bytes
    };
    if raw.len() != 32 {
        return Err(format!(
            "Invalid public key length: expected 32 bytes (Ed25519), got {} after decoding.",
            raw.len()
        ));
    }
    let mut arr = [0u8; 32];
    arr.copy_from_slice(&raw);
    Ok(arr)
}

/// Decode multibase-encoded Ed25519 private key to 32 raw bytes.
pub fn decode_private_key(encoded: &str) -> Result<[u8; 32], String> {
    let bytes = decode_multibase(encoded)?;
    let raw: Vec<u8> = if bytes.len() == 34
        && bytes[0] == ED25519_PRIV_MULTICODEC_PREFIX[0]
        && bytes[1] == ED25519_PRIV_MULTICODEC_PREFIX[1]
    {
        bytes[2..].to_vec()
    } else {
        bytes
    };
    if raw.len() != 32 {
        return Err(format!(
            "Invalid private key length: expected 32 bytes (Ed25519), got {} after decoding.",
            raw.len()
        ));
    }
    let mut arr = [0u8; 32];
    arr.copy_from_slice(&raw);
    Ok(arr)
}

/// Encode 32-byte Ed25519 private key seed as multibase with multicodec prefix.
pub fn encode_private_key(raw: &[u8; 32]) -> String {
    let mut prefixed = Vec::with_capacity(ED25519_PRIV_MULTICODEC_PREFIX.len() + 32);
    prefixed.extend_from_slice(&ED25519_PRIV_MULTICODEC_PREFIX);
    prefixed.extend_from_slice(raw);
    encode_multibase(&prefixed)
}

/// Derive did:syr identifier from Ed25519 public key.
pub fn derive_did(public_key: &[u8; 32]) -> String {
    let mut prefixed = Vec::with_capacity(ED25519_MULTICODEC_PREFIX.len() + 32);
    prefixed.extend_from_slice(&ED25519_MULTICODEC_PREFIX);
    prefixed.extend_from_slice(public_key);
    format!("did:syr:{}", encode_multibase(&prefixed))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encode_decode_multibase_roundtrip() {
        let bytes: Vec<u8> = (0..32).collect();
        let encoded = encode_multibase(&bytes);
        assert!(encoded.starts_with('z'));
        let decoded = decode_multibase(&encoded).unwrap();
        assert_eq!(decoded, bytes);
    }

    #[test]
    fn decode_multibase_rejects_empty() {
        assert!(decode_multibase("").is_err());
    }

    #[test]
    fn decode_multibase_rejects_wrong_prefix() {
        assert!(decode_multibase("mbase58").is_err());
    }

    #[test]
    fn derive_did_format() {
        let pk: [u8; 32] = [1; 32];
        let did = derive_did(&pk);
        assert!(did.starts_with("did:syr:z"));
    }

    #[test]
    fn encode_decode_private_key_roundtrip() {
        let (_, priv_k) = crate::keys::generate_root_keypair();
        let encoded = encode_private_key(&priv_k);
        assert!(encoded.starts_with('z'));
        let decoded = decode_private_key(&encoded).unwrap();
        assert_eq!(decoded, priv_k);
    }

    #[test]
    fn derive_did_decode_roundtrip() {
        let (pub_k, _) = crate::keys::generate_root_keypair();
        let did = derive_did(&pub_k);
        let id_part = did.strip_prefix("did:syr:").unwrap();
        let decoded = decode_public_key(id_part).unwrap();
        assert_eq!(decoded, pub_k);
    }
}
